pkg/lib/synthesize/synthesis.go

```diff
@@ -0,0 +1,696 @@
+			} else {
+				ownedAPIServices = append(ownedAPIServices, v1alpha1.APIServiceDescription{
+					Name:    service.GetName(),
+					Group:   apiGVKN.Group,
+					Version: apiGVKN.Version,
+					Kind:    apiGVKN.Kind,
+				})
+			}
+		}
+	}
+
+	// Gets every APIService included in the bundle and check its GVK against NonInferableCSV to see if a descriptor is
+	// provided. Add the descriptor if it is provided, else record GVKN. Skip if descriptor is provided as a CRD.
+	apiServices, err := s.getAPIServices()
+	if err != nil && err != ObjectTypeNotFoundInBundleError {
+		return nil, err
+	}
+
+	if apiServices != nil {
+		providedApiMap := make(map[v1.GroupVersionKind]struct{})
+		for _, apiService := range *apiServices {
+			providedApiMap[v1.GroupVersionKind{Group: apiService.Spec.Group, Version: apiService.Spec.Version}] = struct{}{}
+		}
+		for _, apiDes := range s.apiServiceDescriptionMap {
+			gv := v1.GroupVersionKind{Group: apiDes.Group, Version: apiDes.Version}
+			if _, ok := providedApiMap[gv]; ok {
+				ownedAPIServices = append(ownedAPIServices, *apiDes)
+				delete(providedApiMap, gv)
+			}
+		}
+		if len(providedApiMap) == 0 {
+			s.apiServiceDescriptionMap = map[v1.GroupVersionKind]*v1alpha1.APIServiceDescription{}
+		}
+	}
+
+	// Everything in apiServiceDescriptionMap either has to be specified in dependency.yaml as required or provided in
+	// the bundle.
+	if len(s.customResourceDefinitionMap) > 0 {
+		return nil, fmt.Errorf("not all descriptors are matched with provided or required APIServices or Service, "+
+			"including: %v", s.customResourceDefinitionMap)
+	}
+
+	return &v1alpha1.APIServiceDefinitions{
+		Owned:    ownedAPIServices,
+		Required: requiredAPIServices,
+	}, nil
+}
+
+func (s *synthesize) getCRDs() (*[]extv1.CustomResourceDefinition, error) {
+	if crdsIF, ok := s.structuredObjectMap[CRDKind]; ok {
+		var crds []extv1.CustomResourceDefinition
+		for _, bufObj := range *crdsIF {
+			crd := extv1.CustomResourceDefinition{}
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(bufObj.UnstructuredContent(), &crd); err != nil {
+				return nil, fmt.Errorf("error decoding %v as %s kind, %v", bufObj.GroupVersionKind(), bufObj.GetKind(), err)
+			}
+			crds = append(crds, crd)
+		}
+		return &crds, nil
+	}
+	return nil, ObjectTypeNotFoundInBundleError
+}
+
+func (s *synthesize) getClusterRoles() (*[]rv1.ClusterRole, error) {
+	if crIF, ok := s.structuredObjectMap[ClusterRoleKind]; ok {
+		var crs []rv1.ClusterRole
+		for _, bufObj := range *crIF {
+			cr := rv1.ClusterRole{}
+			if err := runtime.DefaultUnstructuredConverter.FromUnstructured(bufObj.UnstructuredContent(), &cr); err != nil {
+				return nil, fmt.Errorf("error decoding %v as %s kind, %v", bufObj.GroupVersionKind(), bufObj.GetKind(), err)
+			}
+			crs = append(crs, cr)
+		}
+		return &crs, nil
+	}
+	return nil, ObjectTypeNotFoundInBundleError
+}
+
+func (s *synthesize) getClusterRoleBindings() (*[]rv1.ClusterRoleBinding, error) {
```

Comment:
It strikes me that we should not just support 1 version for every group. We might want to look into ways to do this without a lot of copy/paste.

Reply:
You mean we will support v1, alpha, beta and all? Make sense. Uhm. Maybe this can be future work to define a type for each of the kinds and include the versions we support for them? 

At the moment, since we are not supporting any multiple versions of the same kinds, I wonder if this can be a future task that we keep track of.

Reply:
> At the moment, since we are not supporting any multiple versions of the same kinds, I wonder if this can be a future task that we keep track of.

We need to support both v1 and v1beta1 for CRDs, so I don't think this is something we can punt.

Reply:
We can accept all versions of APIService and CRD since we just need their GVKs. This is reflected in the new commit.

However, for other resources such as (cluster)role, rolebindings, deployments it should not be the case. The operator/v1alpha1.CSV only accepts one version ie: StrategyDeploymentPermissions.rule is rbac/v1 where as accepting rbac/v1beata1.role leads to v1beata1 rule which requires conversion. Same goes to rolebinding as its rolereference, and deployment and its deployment.spec.

Reply:
That's a good point. We're bounded by the versions of the CSV for certain resources for now.